polar-core/src/vm.rs

```diff
@@ -1926,6 +1931,74 @@ impl PolarVirtualMachine {
         Ok(QueryEvent::None)
     }
 
+    /// Check for partially bound or unbound variable arguments to an external call.
+    /// If any arguments are partially bound or unbound, return an error.
+    /// The only exception is the method `OsoRoles.role_allows()`, which expects a partially bound resource argument
+    fn check_partial_args(
+        &mut self,
+        object: &Term,
+        field: &Term,
+        value: &Term,
+    ) -> PolarResult<bool> {
+        // If the lookup is a `Value::Call`, then we need to check for partial args
+        let mut has_partial_arg = false;
+        if let Value::Call(Call { name, args, kwargs }) = field.value() {
+            // get all arg values (args + kwargs)
+            let mut arg_values = args.clone();
+            if let Some(kwarg_map) = kwargs.clone() {
+                let mut v: Vec<Term> = kwarg_map.values().cloned().collect();
+                arg_values.append(&mut v)
+            }
+            for (i, arg) in arg_values.iter().enumerate() {
+                // Partial/unbound variables cannot be passed to external methods from Polar
+                if let Value::Variable(v) = arg.value() {
+                    match self.binding_manager.variable_state(v) {
+                        // bound variables are fine, continue
+                        VariableState::Bound(_) => continue,
+                        // partial variables are only fine if being passed into "role_allows"
+                        VariableState::Partial => {
+                            if let Value::ExternalInstance(external) =
+                                self.deep_deref(&object).value()
+                            {
+                                if let Some(repr) = external.repr.clone() {
+                                    // only allow the third argument (resource) to be partially bound
+                                    if repr.contains("sqlalchemy_oso.roles2.OsoRoles")
+                                        && name.0 == "role_allows"
+                                        && i == 2
```

Comment:
This `i == 2` check is not correct in the presence of keyword arguments.

Reply:
do we want it to accept kwargs?

Reply:
Actually don't think this is true... why shouldn't it be correct if kwargs?

Reply:
The keyword arguments are unordered.  You mean to check that the parameter either in position 2 or with the name `resource` is partial. However, if someone uses all keyword arguments, there is not a guarantee that the `resource `argument is in position 2.

This function would be great to test.

Reply:
Ohhhh right sorry was misreading